compiler/rustc_passes/src/dead.rs

```diff
@@ -677,57 +686,178 @@ impl<'tcx> DeadVisitor<'tcx> {
-                let descr = self.tcx.def_kind(def_id).descr(def_id.to_def_id());
-                let mut err = lint.build(&format!("{} is never {}: `{}`", descr, participle, name));
-                let hir = self.tcx.hir();
-                if let Some(encl_scope) = hir.get_enclosing_scope(id)
-                    && let Some(encl_def_id) = hir.opt_local_def_id(encl_scope)
-                    && let Some(ign_traits) = self.ignored_derived_traits.get(&encl_def_id)
-                {
-                    let traits_str = ign_traits
-                        .iter()
-                        .map(|(trait_id, _)| format!("`{}`", self.tcx.item_name(*trait_id)))
-                        .collect::<Vec<_>>()
-                        .join(" and ");
-                    let plural_s = pluralize!(ign_traits.len());
-                    let article = if ign_traits.len() > 1 { "" } else { "a " };
-                    let is_are = if ign_traits.len() > 1 { "these are" } else { "this is" };
-                    let msg = format!(
-                        "`{}` has {}derived impl{} for the trait{} {}, but {} \
-                        intentionally ignored during dead code analysis",
-                        self.tcx.item_name(encl_def_id.to_def_id()),
-                        article,
-                        plural_s,
-                        plural_s,
-                        traits_str,
-                        is_are
-                    );
-                    err.note(&msg);
-                }
-                err.emit();
+        let field_or_variant_local_def_id = |parent_node, node| -> Option<(LocalDefId, bool)> {
+            match (parent_node, node) {
+                // struct/union fields
+                (
+                    Node::Item(hir::Item {
+                        kind:
+                            hir::ItemKind::Struct(hir::VariantData::Struct(..), ..)
+                            | hir::ItemKind::Union(hir::VariantData::Struct(..), ..),
+                        def_id,
+                        ..
+                    }),
+                    Node::Field(..),
+                ) => Some((*def_id, true)),
+                // enum variants
+                (
+                    Node::Item(hir::Item { kind: hir::ItemKind::Enum(..), def_id, .. }),
+                    Node::Variant(..),
+                ) => Some((*def_id, false)),
+                // variant fields
+                (
+                    Node::Variant(hir::Variant { data: hir::VariantData::Struct(..), id, .. }, ..),
+                    Node::Field(..),
+                ) => Some((id.owner, true)),
+                _ => None,
+            }
+        };
+        if let Some(parent_hir_id) = self.tcx.hir().find_parent_node(id)
+            && let Some(parent_node) = self.tcx.hir().find(parent_hir_id)
+            && let Some(node) = self.tcx.hir().find(id)
+            && let Some((did, is_field)) = field_or_variant_local_def_id(parent_node, node)
```

Comment:
Instead of doing this, why don't we modify the visitor to *not* call this function from the variant and field visits, but rather from the type visitor so that we *don't* need to look for the parent item, we pass it in.